### DISCUSSION TITLE
Data to Dataloader

### DISCUSSION THREAD

**Discussion by Chrisable** [pyg-team/pytorch_geometric#5236] [General]
I have data of Data type.How to do dataloader?
Data:
`Data(x=[23974, 12, 207, 2], edge_index=[2, 1722], edge_attr=[1722], y=[23974, 12, 207, 2], num_nodes=207)`

**Comment by rusty1s** [comment]
It depends on what your `data` represents? What is the data `x` holds and what does it shape represents (since it does not follow the usual convention of `[num_nodes, num_features]`)?

**Reply by Chrisable** [reply]
x: (num_samples, input_length, num_nodes, input_dim)

**Reply by saiden89** [reply]
Can you please provide a description of for each element in `x`? Specifically, what is the difference between `num_nodes` and `num_samples`? And between `input_length` and `input_dim`?

**Reply by rusty1s** [reply]
It looks like you have `num_samples` many examples, and for each graph a sequence of features (`input_length`). In that case, you could batch as follows:
```python
data_list = []
for i in range(data.x.size(0):
    sub_data = copy.copy(data)
    sub_data.x = sub_data.x[i].permute(1, 0, 2)
    sub_data.y = sub_data.y[i].permute(1, 0, 2)
    data_list.append(sub_data)

loader = DataLoader(data_list, batch_size=32, ...)
```

**Reply by Chrisable** [reply]
Thank you very much !!!

**Reply by Chrisable** [reply]
How can such data be divided into subgraphs so that each subgraph can train a model, then aggregate the models of each subgraph, and finally do the downstream task？

**Reply by Chrisable** [reply]
@rusty1s

**Reply by rusty1s** [reply]
Does that mean you want to train an isolated model for each subgraph? In that case, you would need to initialize a model for each `data` object in `data_list`. I am not sure if that makes sense though - let me know if I am missing something.

**Reply by Chrisable** [reply]
`x: (num_samples, input_length, num_nodes, input_dim)  y: (num_samples, output_length, num_nodes, output_dim) `
I plan to divide the training set into several subgraphs, then train a model for each subgraph.Then input the validation set into each subgraph model, and then aggregate the output of each subgraph model. However, the **num_nodes** dimension of the verification set is still the **num_nodes** dimension before the subgraph is divided, and the **num_nodes** dimension of each subgraph is the **num_nodes** dimension after the partition, so my idea is not feasible at present.

**Reply by rusty1s** [reply]
Can you clarify? Sorry :( I think you should be able to split your data into subgraphs during training, and just train on individual subgraphs. For inference though, I suggest that you perform it on the full-graph. This way, you do not need any aggregation logic.

**Reply by Chrisable** [reply]
But  when i split the data into subgraphs during training,each input of subgraphs is (num_samples, input_length, **sub_num_nodes**, input_dim),for example:(2000,12,100,2).
For inference,the input  is (1000,12,200,2),how to input validation set into subgraph model?

**Reply by rusty1s** [reply]
I think your model should be independent of the number of nodes, right? That is, it shouldn't assume a fixed number of nodes per sample. This way, the number of nodes can vary across training/prediction.

**Reply by Chrisable** [reply]
right

**Comment by Chrisable** [comment]
In traffic prediction, if a traffic network is divided into several partitions in the training stage, each partition trains a model, and the verification set is also changed into corresponding partition according to the division in the training stage. Finally,In the verification phase，each partition has a prediction.So,should these predictions be combined or aggregated to obtain the predicted value of each node  ？

**Reply by rusty1s** [reply]
I think you only need to aggregate predictions across partitions in case these partitions overlap. Is that the case for you?

**Reply by Chrisable** [reply]
If there is no overlap in each partition, is it not necessary to aggregate? If so, the model in each partition will lead to poor results due to the small number of samples.Is there any way？

**Reply by rusty1s** [reply]
I don't think there is a need to aggregate in that case. I am not an expert on the task you want to solve, but you could try to increase the partition size during inference in order to let the model know about a larger number of samples.

**Reply by Chrisable** [reply]
Thank you
